Resource compiler

Hazardu opened the pull request:
The resources added are defined in CMakeLists.txt inside src/irr/resources

Hazardu committed:
Resources

Hazardu committed:
Delete embedResource.c

devshgraphicsprogramming commented on CMakeLists.txt:
```diff
@@ -35,6 +35,10 @@ if (IRR_COMPILE_WITH_CUDA)
 
+#resource embedding
+if(_IRR_RESOURCE_COMPILE_EVEN_IN_DEBUG)
+	add_subdirectory(src/irr/resource)
```
make it `src/irr/builtin`

devshgraphicsprogramming replied:
make it `src/irr/builtin`

devshgraphicsprogramming replied:
make it `src/irr/builtin`

devshgraphicsprogramming commented on src/irr/resource/embedResource.py:
```diff
@@ -0,0 +1,29 @@
+    outp = open(outputfilename,"w+")
+
+    outp.write("#include <stdlib.h>\n")
+    outp.write("const uint8_t* %s[] = {\n" % name)
```
needs to be a sized array
`const uint8_t %s[%d]` with %d giving the size

devshgraphicsprogramming commented on src/irr/resource/exampleResource.txt:
```diff
@@ -0,0 +1 @@
+aaaaaaaaaaaaaaaaaaaaaaaaaaab
```
make an example out of the built-in shaders (for that you need to work off the `shader_pipeline` branch)

devshgraphicsprogramming commented on CMakeLists.txt:
```diff
@@ -35,6 +35,10 @@ if (IRR_COMPILE_WITH_CUDA)
 endif()
 
+#resource embedding
+if(_IRR_RESOURCE_COMPILE_EVEN_IN_DEBUG)
```
you need to declare this option with a proper default, also I'd call it `IRR_EMBED_BUILTIN_RESOURCES`

it also needs to be defined in the installed configure header in `include/irr/config` so the C++ code can check the macro `_IRR_EMBED_BUILTIN_RESOURCES` and decide whether to load from embedded `uint8_t`  array or some file in `include/irr/builtin/`

Hazardu committed:
still need to figure out what to do in src/irr/builtin/CMakeLists.txt

devshgraphicsprogramming commented on src/irr/builtin/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+cmake_minimum_required(VERSION 3.9)
```
no need for this, its in the top level cmake

devshgraphicsprogramming replied:
?

devshgraphicsprogramming commented on src/irr/builtin/shader1_res.h:
```diff
@@ -0,0 +1,41 @@
+#include <stdlib.h>
```
don't commit the resource files.

devshgraphicsprogramming replied:
?

devshgraphicsprogramming commented on src/irr/builtin/embedResource.py:
```diff
@@ -0,0 +1,51 @@
+    #opening a file
+    outp = open(outputfilename,"w+")
+
+    outp.write("#include <stdlib.h>\n")
```
you need to embed the arrays in an appropriate namespace
```cpp
namespace irr
{
namespace builtin
{
```

devshgraphicsprogramming commented on src/irr/builtin/embedResource.py:
```diff
@@ -0,0 +1,51 @@
+    outp.write("#include <stdlib.h>\n")
+
+    # outp.write("const uint8_t* resource[%d] = {\n" % fsize) # array size is not required
+    outp.write("const uint8_t* resource[] = {\n")
```
you need a unique variable name instead of `resource`

I'd take the original virtual path like `irr/builtin/X/Y/Z/something.type` and I would replace the `/` or `\` and `.` with `_` and use that as a name, for now

devshgraphicsprogramming commented on src/irr/builtin/shader1.txt:
```diff
@@ -0,0 +1,10 @@
+#https://github.com/Crisspl/IrrlichtBAW/blob/shaders/src/irr/asset/IAssetManager.cpp#L201        
```
give this file the same name as its virtual path.

devshgraphicsprogramming commented on src/irr/builtin/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+function(irr_add_resource filename)
+  message(STATUS "Starting generating resource from ${filename}")
+  add_custom_command(
+    OUTPUT ${CMAKE_CURRENT_SOURCE_DIR}/${filename}.h
```
you don't output to source dir, you output to binary/build dir

devshgraphicsprogramming replied:
not sure you should be using the `CMAKE_CURRENT_SOURCE` variable

devshgraphicsprogramming commented on include/irr/config/BuildConfigOptions.h.in:
```diff
@@ -21,5 +21,6 @@
 
 // extra config
 #cmakedefine __IRR_FAST_MATH
+#cmakedefine _IRR_EMBED_BUILTIN_RESOURCES
```
you're missing the magical glue that defines it for this config header (should be in `./source/Irrlicht/CMakeLists.txt`)

Hazardu committed:
namespace irr/builtin added to embedded resources

Hazardu committed:
Resources split into a header and cpp file

devshgraphicsprogramming commented on src/irr/builtin/builtinDataGen.py:
```diff
@@ -0,0 +1,54 @@
+    #opening a file
+    outp = open(outputFilename,"w+")
+  
+    outp.write("#include irr/builtin/builtinResources.h\n\n")
```
missing quotes

Hazardu committed:
Added missing quotes

Hazardu committed:
I need help, neither is appearing to be working

 COMMAND python nor COMMAND ${PYTHON_EXECUTABLE} invoke the scripts, but inside src\irr\builtin\CMakeFiles files are being created

devshgraphicsprogramming commented on src/irr/builtin/CMakeLists.txt:
```diff
@@ -2,22 +2,22 @@ function(irr_add_builtin_resource)
 
   # set up dependency
-  add_custom_target(builtin_resources DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/build/builtinResourceData.cpp ${CMAKE_CURRENT_BINARY_DIR}/build/builtinResources.h)
+  add_custom_target(builtin_resources DEPENDS "${CMAKE_CURRENT_BINARY_DIR}/builtinResourceData.cpp" "${CMAKE_CURRENT_BINARY_DIR}/builtinResources.h")
```
check if you have the paths right, you can use cmake `message` to check what `${CMAKE_CURRENT_BINARY_DIR}/builtin` evalutates to

Hazardu committed:
Finished resource compiler

Hazardu committed:
Changed DIR of generated files to source folder

Hazardu committed:
Criss' fixes

devshgraphicsprogramming commented on src/irr/builtin/builtinshader.cpp:
```diff
@@ -0,0 +1,11 @@
+#include "irr/builtin/resuorces.h"
```
don't commit these

devshgraphicsprogramming commented on src/irr/builtin/CMakeLists.txt:
```diff
@@ -1,19 +1,33 @@
 
 if(IRR_EMBED_BUILTIN_RESOURCES)
-	irr_add_resource(shader1.txt embeddedShader)
+  #fill this list with resource paths relative to top level of IrrlichtBaW's source
+  set(BUILTIN_RESOURCE_LIST src/irr/builtin/shader1.txt)
+
+	irr_add_builtin_resource()
```
this has really bad syntax

devshgraphicsprogramming replied:
you should be passing the list to the function, not using a weird global cmake variable

devshgraphicsprogramming commented on src/irr/builtin/CMakeLists.txt:
```diff
@@ -1,19 +1,33 @@
 if(IRR_EMBED_BUILTIN_RESOURCES)
-	irr_add_resource(shader1.txt embeddedShader)
+  #fill this list with resource paths relative to top level of IrrlichtBaW's source
+  set(BUILTIN_RESOURCE_LIST src/irr/builtin/shader1.txt)
```
why is `BUILTIN_RESOURCE_LIST` some sort of global variable?

devshgraphicsprogramming commented on src/irr/builtin/CMakeLists.txt:
```diff
@@ -2,19 +2,17 @@ function(irr_add_builtin_resource)
-    COMMAND python ${CMAKE_CURRENT_SOURCE_DIR}/buitlinHeaderGen.py ${CMAKE_CURRENT_BINARY_DIR}/build/builtinResources.h ${CMAKE_SOURCE_DIR} ${BUILTIN_RESOURCE_LIST}
-    COMMAND ${PYTHON_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/builtinDataGen.py ${CMAKE_CURRENT_BINARY_DIR}/build/builtinResourceData.cpp ${CMAKE_SOURCE_DIR} ${BUILTIN_RESOURCE_LIST}
+    OUTPUT "${CMAKE_CURRENT_BUILD_DIR}/builtinResources.h" "${CMAKE_CURRENT_BUILD_DIR}/builtinResourceData.cpp"
+    COMMAND python ${CMAKE_CURRENT_SOURCE_DIR}/builtinHeaderGen.py ${CMAKE_CURRENT_BUILD_DIR}/builtinResources.h ${CMAKE_SOURCE_DIR} ${BUILTIN_RESOURCE_LIST}
```
`${PYTHON_EXECUTABLE}` is the correct one to use

devshgraphicsprogramming commented on src/irr/builtin/CMakeLists.txt:
```diff
@@ -2,15 +2,15 @@ function(irr_add_builtin_resource)
-    OUTPUT "${CMAKE_CURRENT_BUILD_DIR}/builtinResources.h" "${CMAKE_CURRENT_BUILD_DIR}/builtinResourceData.cpp"
-    COMMAND python ${CMAKE_CURRENT_SOURCE_DIR}/builtinHeaderGen.py ${CMAKE_CURRENT_BUILD_DIR}/builtinResources.h ${CMAKE_SOURCE_DIR} ${BUILTIN_RESOURCE_LIST}
-    COMMAND ${PYTHON_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/builtinDataGen.py ${CMAKE_CURRENT_BUILD_DIR}/builtinResourceData.cpp ${CMAKE_SOURCE_DIR} ${BUILTIN_RESOURCE_LIST}
+    OUTPUT "${CMAKE_CURRENT_SOURCE_DIR}/builtinResources.h" "${CMAKE_CURRENT_SOURCE_DIR}/builtinResourceData.cpp"
```
WHY!?

Hazardu committed:
back to binary dir

devshgraphicsprogramming commented on include/CConcurrentObjectCache.h:
```diff
@@ -4,184 +4,184 @@
 #include "CObjectCache.h"
```
can you revert the change to this file?

devshgraphicsprogramming commented on src/irr/asset/IAssetManager.cpp:
```diff
@@ -64,18 +64,9 @@
 #endif
-
-#include "irr/asset/CGLSLLoader.h"
-#include "irr/asset/CSPVLoader.h"
```
why did you remove the extra headers?

devshgraphicsprogramming commented on examples_tests/03.GPU_Mesh/main.cpp:
```diff
@@ -4,15 +4,35 @@
-//#include "../../ext/ScreenShot/ScreenShot.h"
-
+#include "../../ext/ScreenShot/ScreenShot.h"
+#include "../../src/irr/builtin/builtinResources.h"
```
don't touch my samples, try to embed one of the shaders in `IAssetManager` instead of having it as a rawstring and test (example 35 uses builtin shaders)

Hazardu committed:
Merge branch 'shader_pipeline' into resource_compiler

devshgraphicsprogramming commented on include/irr/config/BuildConfigOptions.h.in:
```diff
@@ -18,12 +18,30 @@
 #cmakedefine _IRR_COMPILE_WITH_OPENGL_
 #cmakedefine _IRR_COMPILE_WITH_VULKAN_
 
-// compute interop
```
WHAT THE FUCK ARE YOU DOING?
